rfhub2/cli/api_client.py

```diff
@@ -54,7 +54,7 @@ def _post_request(self, endpoint: str, data: Dict) -> post:
         Sends post request to collections or keywords endpoint.
         """
         request = self.session.post(url=f'{self.api_url}/{endpoint}/', json=data)
-        return request.json()
+        return request.status_code, request.json()
```

Comment:
Doing it that way is definitely cleaner, should I add it to _get and _delete methods as well to keep consistency?

Reply:
i would not expose the status code outside, and handle the auth and deserialization issues in the client. That was the initial idea for returning the pydantic model instances here and having Collection/Keyword as returned type and not just Dict. For example first you would check the status code and raise some AuthenticationError if 401 and then try to create and return pydantic model instance out of request.json() or raise DeserializationError otherwise. Then when you use the client methods you could rely on instances of Collection/Keyword object having all the required fields present and validated. And you could catch errors raised by client to reraise them or raise StopIteration, etc. without the need to add any other logic related to request and request body schema outside of the client.

Reply:
So if I understood You correctly:

1. Type annotation for this method should stay the same, but type annotation for `create_collection` and `create_keyword` should be changed to Collection and Keyword.
2. Additional logic to handle status code 401 and throw AuthenticationError.
3. Additional logic to handle Deserialization - handle status code 422 in `_post_request`.
4. Pass request.json() to `create_collection` and `create_keyword` and deserialize json to either Collection or Keyword model.
5. Change all methods to work like that.

> This issue involves adding definition of our domain specific errors and using pydantic on cli side so if you want we can leave it like this for now and refactor it in the subsequent PR.

I agree, let's have initial working version and then refactor it step by step in small PR.

Reply:
1. Current type annotation should be updated - i added separate comment about it. After changes described above return value will be Collection or Keyword. method responsible for mapping dict to model instance will be called inside this method.
3. It should be both Serialization and Deserialization error - serialization is the one you described and deserialization will be when api will respond with 201, but client won't be able to create model instance from returned json.